modules/generator/storage/instance.go

```diff
@@ -0,0 +1,111 @@
+package storage
+
+import (
+	"context"
+	"errors"
+	"os"
+	"path/filepath"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	prometheus_config "github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/scrape"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/storage/remote"
+	"github.com/prometheus/prometheus/tsdb/agent"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+)
+
+type Storage interface {
+	storage.Appendable
+
+	// Close closes the storage and all its underlying resources.
+	Close() error
+}
+
+type storageImpl struct {
+	walDir        string
+	wal           *agent.DB
+	remoteStorage *remote.Storage
+
+	logger log.Logger
+}
+
+var _ Storage = (*storageImpl)(nil)
+
+// New creates a metrics WAL that remote writes its data.
+func New(cfg *Config, tenant string, reg prometheus.Registerer, logger log.Logger) (Storage, error) {
+	logger = log.With(logger, "tenant", tenant)
+	reg = prometheus.WrapRegistererWith(prometheus.Labels{"tenant": tenant}, reg)
+
+	walDir := filepath.Join(cfg.Path, tenant)
+
+	level.Info(logger).Log("msg", "creating WAL", "dir", walDir)
+
+	// Create WAL directory with necessary permissions
+	err := os.MkdirAll(walDir, 0o755)
+	if err != nil {
+		level.Warn(logger).Log("msg", "could not create directory for metrics WAL", "err", err)
+	}
```

Comment:
Isn't this redundant with lines 54-57 creating the same directory + `/wal`?

Reply:
Yes, that's correct I think. I started out with the first `MkdirAll` and added the second one later to avoid the error from remote storage trying to read `<walDir>/<tenant>/wal`.

MkdirAll says:

> _The permission bits perm (before umask) are used for all directories that MkdirAll creates._

So the second `MkdirAll` should also create the tenant directory with the right permission bits.